### ISSUE TITLE
Please provide a MATE-specific implementation of xdg-desktop-portal

### ISSUE THREAD

**Issue by nielsdg** [mate-desktop/mate-desktop#504]
#### Expected behaviour

[XDG desktop portals](https://github.com/flatpak/xdg-desktop-portal) are a set of D-Bus APIs that work independent of DE, display server, and works in containerized applications like Flatpak, Snap and others. It is already implemented by GNOME, KDE, Elementary and wlroots. It provides for example a way of grabbing the screen that can only happen after a user consents to it.

GTK-based DEs already have a shared implementation for a subset of portals, called [xdg-desktop-portal-gtk](https://github.com/flatpak/xdg-desktop-portal-gtk). However, certain APIs are compositor-specific and can't be put there. GNOME has [xdg-desktop-portal-gnome](https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome) while Elementary also [has their own version](https://github.com/elementary/portals).

The expectation is that more and more applications migrate to these APIs, as they provide cross-platform support. Certain display protocols (like Wayland) also explicitly don't provide such API.

#### Actual behaviour

Using the [PickColor](https://flatpak.github.io/xdg-desktop-portal/#gdbus-method-org-freedesktop-portal-Screenshot.PickColor) method of the Screenshot portal doesn't work in Mate, as it's not implemented.

#### Steps to reproduce the behaviour

GIMP 2.10.30 switched to using the XDG-desktop-portal, so now it's color picker is broken on MATE.

#### MATE general version

Latest stable

#### Package version

Latest stable

#### Linux Distribution

* Reported on the GIMP issue tracker with Linux Mint 20.1 MATE, but pretty sure I can reproduce it everywhere

#### Link to bugreport of your Distribution (requirement)

I'm fairly sure this is distro-independent. If you really want me to, I can make a downstream bug

**Comment by lukefromdc** [comment]
I just updated to gimp 2.10.30 as a systemwide install from a .deb package and the color picker seems to work same as before
on Debian Unstable. I do not have support for flatpacks or snaps installed so cannot test those

**Comment by nielsdg** [comment]
I guess that's a bit thanks to some crude checking in GIMP if XDG desktop portal is working or not: we check if the XDG service exists on the session bus, and is pingable over D-Bus. If that's the case, then we say it's good to go; if it's not, we fall back to X11 color picking.

The problem is then that anyone who has xdg-desktop-portal-gtk (or something that depends on it, like xdg-desktop-portal-gnome or xdg-desktop-portal-elementary) installed will see that the XDG portal D-Bus service is alive and attempt to do color picking/screenshotting using that method, which errors out.

**Comment by lukefromdc** [comment]
That sounds likely: I use a very old, continuously updated system and have never
chosen to use some of the newer OS level changes on my own system. Any time
I have to "nuke from orbit" I have plenty of rollback images of the whole root FS.
Thus no Flatpack/Snap support or their underlying libs and daemons separate from
what a 2012 era system would have.

**Comment by nielsdg** [comment]
FWIW, I don't have the time to implement this, but I can absolutely support anyone who's willing to do so. Certain things shouldn't be too hard anyway (for example: the Screenshot portal can be partially implemented using the `mate-screenshot` binary)

**Comment by stephematician** [comment]
Would it be simpler to adapt the GNOME desktop portal to MATE?

**Comment by nielsdg** [comment]
The GNOME desktop portal can be used as inspiration for a MATE-portal, but adapting it doesn't really make sense, as it's bound to the internal APIs and release cadence of GNOME, not that of MATE.

It might be interesting to mention here that the folks over at Cinnamon have made https://github.com/linuxmint/xdg-desktop-portal-xapp, so this might already be resolved? I'm not sure how much the MATE developers have been involved in this one.

**Comment by raveit65** [comment]
Nope, xapps are cinamon only production. There xapps are only forks from origin MATE applications like Atril, eom, engrampa. In other words we don't need that.
Writing a xdg-desktop-portal-mate would make sense.
But we are a very small team , so voluntary are wanted for this work.

**Comment by smcv** [comment]
See also mate-desktop/mate-session-manager#311: adding a `mate-portals.conf` would be a small step towards this, and very straightforward to implement (add and install one small text file which codifies your preferred x-d-p backends).

> xapps are cinamon only production. There xapps are only forks from origin MATE applications like Atril, eom, engrampa. In other words we don't need that.

The implementation of x-d-p-xapp seems to disagree with this, and is targeting Cinnamon, MATE and XFCE, not just Cinnamon - it has `UseIn=X-Cinnamon;MATE;XFCE;`. This is one of the reasons why x-d-p added the `portals.conf(5)` mechanism as a replacement for `UseIn` - so that it can be the desktop environment maintainer, not the portal backend implementor, that makes the choice of which portal backend(s) are the most appropriate to use by default in each desktop.

I don't normally run any of those desktop environments myself, but x-d-p-xapp seems to include MATE-specific implementation code for at least some interfaces - although notably not the screenshot interface (so it will not solve the problem with GIMP that was originally reported here).

**Comment by smcv** [comment]
> Basically mate-desktop is the main package to provide configs for Mate desktop

If that's the policy, then adding a mate-portals.conf in mate-desktop might be more appropriate than in mate-session-manager. Are you able to move https://github.com/mate-desktop/mate-session-manager#311 to here? (I can't seem to do that as the issue submitter, so I think only a maintainer will be able to do that.)

mate-portals.conf should probably be in the same place where you'd put a mate-mimeapps.list, if you have one - they're very similar mechanisms.

> Again, again xapps was a idea from linuxmint devs and we never agreed to them to do that for general linux/unix

The good news is that with x-d-p >= 1.17, this is under the desktop environment's control, not the portal backend's: if Mate installs a mate-portals.conf, it can prevent x-d-p-xapp from being used under Mate. For older x-d-p versions, if Mate developers would like x-d-p-xapp to stop declaring itself as suitable for use in Mate, please open an issue there - as a non-user of Mate, it doesn't seem right for me to do that.

> Keep in mind that all rpm based distros don't use that in default Mate installations. Not sure about ubuntu Mate and debian Mate.

In at least Debian, it's straightforward to install more than one desktop environment in parallel, which could result in x-d-p-xapp being pulled in as a result of installing Cinnamon or XFCE (even if Mate doesn't have a dependency on it). Debian 12 didn't include x-d-p-xapp, but Debian 13 will. However, Debian 13 will also include x-d-p >= 1.17, so Mate will be able to use a mate-portals-conf to prevent x-d-p-xapp from being used unintentionally.

**Comment by clefebvre** [comment]
> It might be interesting to mention here that the folks over at Cinnamon have made https://github.com/linuxmint/xdg-desktop-portal-xapp, so this might already be resolved? I'm not sure how much the MATE developers have been involved in this one.

It is resolved. This bug is fixed for all 3 DEs just by installing xdg-desktop-portal-xapp. It's designed to work and to support MATE, Cinnamon and Xfce.

> Nope, xapps are cinamon only production. 

This makes no sense. It's the complete opposite: XApp is specifically designed for things which are common to MATE/Cinnamon/Xfce. It's its goal. If something's only made for Cinnamon it's not an Xapp project, it's a Cinnamon project.

> There xapps are only forks from origin MATE applications like Atril, eom, engrampa. In other words we don't need that.

Not really. One Xapp subproject called Xreader is indeed forked from Atril. None are forked from eom nor engrampa. Whether you need cross-DE apps which work well in all DEs instead of maintaining your own for no good reason, that's your call. In my opinion yes, you do need that.

> Writing a xdg-desktop-portal-mate would make sense.

It's already done so no, it would make no sense.

> But we are a very small team , so voluntaries are wanted for this work.

The work is already done. We're all very small teams. When we develop something we want it to work in as many places as possible. The more we go forward the less we design things to be DE specific or distro-specific.

You can decide that you don't want to use it because you don't like the fact that it's made by us, and that's a pity. But that doesn't make it any less compatible with your distro and your desktop environment. It isn't just compatible, it's specifically designed to work for all distros and these 3 desktops.

**Comment by raveit65** [comment]
@nielsdg 
In a fedora 38 VM with gimp-2.10.34 i removed the xdg-desktop-portal complete and i don't have a XDG service anymore with `systemctl --user list-unt-files`. But the color-picker in gimp works perfect.
Are you sure about the gimp version?
Or are you using a self compiled version with using xdg-desktop-portal?

**Comment by clefebvre** [comment]
Removing xdg-desktop-portal works around the issue. It makes Gimp fallback to X11 for screen capturing which fixes the color picker.

**Comment by smcv** [comment]
> Removing xdg-desktop-portal works around the issue. It makes Gimp fallback to X11 for screen capturing which fixes the color picker.

That's consistent with what was said in https://github.com/mate-desktop/mate-desktop/issues/504#issuecomment-1002253152: it seems Gimp isn't checking whether x-d-p has a working implementation of the Screenshot portal, but only checking whether x-d-p is present at all.

**Comment by raveit65** [comment]
Again, with or without xdg-desktop-portal installed or a XDG portal service is running the issue isn't reproducible in fedora 38. The color-picker of gimp works always. 

>  Reported on the GIMP issue tracker with Linux Mint 20.1 MATE, but pretty sure I can reproduce it everywhere

Looks like an linuxmit only issue.

**Comment by clefebvre** [comment]
> > Removing xdg-desktop-portal works around the issue. It makes Gimp fallback to X11 for screen capturing which fixes the color picker.
> 
> That's consistent with what was said in [#504 (comment)](https://github.com/mate-desktop/mate-desktop/issues/504#issuecomment-1002253152): it seems Gimp isn't checking whether x-d-p has a working implementation of the Screenshot portal, but only checking whether x-d-p is present at all.

Yeah that's exactly it. It checked whether or not there was a portal interface and used it if it found one. If it didn't it fell back to X11. xdg-portal-desktop-gtk was designed with GNOME in mind and used GNOME Shell's screen capture dbus interface, which obviously doesn't exist in MATE. That's how it failed to work.

I used the past tense above because when we developed a portal for MATE/Cinnamon/Xfce this bug was known and solutions were planned for both GIMP and xdg-portal-desktop-gtk.

As Neils suggests here, this bug seems to have been fixed in recent versions of GIMP and xdg-desktop-portal-gtk.

![image](https://github.com/mate-desktop/mate-desktop/assets/1138515/69d6c22d-3c82-49be-9a73-0fa3e255dbfa)

> Looks like an linuxmint only issue.

No. It affects any distribution using versions of GIMP and xdg-desktop-portal-gtk lower than the ones mentioned above.
